modelzoo/admin.py

```
@@ -0,0 +1,475 @@
+                        conn,
+                        username=username,
+                        password=password,
+                        token=token,
+                        admin=False,
+                    )
+                    u.parent = self
+                    return u
+        u = UserAuth(conn, username=username, password=password, token=token)
+        return u
+
+    @property
+    def admin(self):
+        if self.adminF:
+            return self.admin_token
+        if self.parent is None:
+            return self.token
+        return self.parent.admin
+
+
+class ModelZooConnection(object):
+    def __init__(
+        self,
+        address: typing.Optional[str] = "grpc.modelzoo.live",
+        token: typing.Optional[str] = "",
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+    ):
+        self.conn = None
+        self.address = address
+        self.conn_error = ModelZooConnectionException(
+            "Must be connected to ModelZoo instance before issuing requests. Please call connect()."
+        )
+        try:
+            self.auth = UserAuth(
+                self, username=username, password=password, token=token
+            )
+        except AuthenticationException as e:
+            self.auth = None
+
+    def connect(self, address: typing.Optional[str] = "grpc.modelzoo.live"):
+        self.address = self.address if address is None else address
+        channel = grpc.insecure_channel(self.address)
+        self.conn = ModelStub(channel)
+
+    def authenticate(
+        self,
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+        token: typing.Optional[str] = "",
+    ) -> None:
+        try:
+            self.auth = UserAuth(
+                self, username=username, password=password, token=token
+            )
+        except InsufficientCredentialsException:
+            raise ModelZooConnectionException(
```

Why don't we just let it throw? User should be able to see InsufficientCredentialsException.

---

Plus the messages are basically the same